File: cpp/355.DesignTwitter.cpp
const int MOST_RECENT_TWEET_NUM = 10;

class Tweet {
public:
    int tweetId;
    int timestamp;

    Tweet(int tweetId, int timestamp) : tweetId(tweetId), timestamp(timestamp) {}
};

bool tweet_cmp(Tweet *a, Tweet *b) {
    return a->timestamp < b->timestamp;
}

class User {
public:
    int id;
    vector<Tweet *> tweets;
    set<User *> following;

    User(int id) : id(id) {}
};

class Twitter {
public:
    /** Initialize your data structure here. */
    Twitter() {
    }

    /** Compose a new tweet. */
    void postTweet(int userId, int tweetId) {
        auto user = getUser(userId);
        user->tweets.emplace_back(new Tweet(tweetId, timestamp++));
    }

    /** Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent. */
    vector<int> getNewsFeed(int userId) {
        set<Tweet *, decltype(tweet_cmp) *> newsFeed(tweet_cmp);
        auto user = getUser(userId);

        for (auto followee:user->following) {
            int tweetNum = followee->tweets.size();
            for (int i = tweetNum - 1; i >= 0 && i >= tweetNum - MOST_RECENT_TWEET_NUM; i--) {
                auto currentTweet = followee->tweets[i];
                if (newsFeed.size() < MOST_RECENT_TWEET_NUM) {
                    newsFeed.insert(currentTweet);
                    continue;
                }

                auto firstTweet = *newsFeed.begin();
                if (firstTweet->timestamp < currentTweet->timestamp) {
                    newsFeed.erase(newsFeed.begin());
                    newsFeed.insert(currentTweet);
                }
            }
        }

        int num = newsFeed.size();
        vector<int> ans(num);
        while (!newsFeed.empty()) {
            ans[--num] = (*newsFeed.begin())->tweetId;
            newsFeed.erase(newsFeed.begin());
        }
        return ans;
    }

    /** Follower follows a followee. If the operation is invalid, it should be a no-op. */
    void follow(int followerId, int followeeId) {
        auto follower = getUser(followerId);
        auto followee = getUser(followeeId);
        follower->following.insert(followee);
    }

    /** Follower unfollows a followee. If the operation is invalid, it should be a no-op. */
    void unfollow(int followerId, int followeeId) {
        if (followerId == followeeId) return;
        if (!checkUserExists(followerId) || !checkUserExists(followeeId)) return;
        auto follower = getUser(followerId);
        auto followee = getUser(followeeId);
        follower->following.erase(followee);
    }

private:
    int timestamp = 0;
    map<int, User *> users;

    User *getUser(int userId) {
        auto ptr = users.find(userId);
        if (ptr == users.end()) {
            auto user = new User(userId);
            user->following.insert(user);
            users.insert({userId, user});
            return user;
        }
        return ptr->second;
    }

    bool checkUserExists(int userId) {
        return users.find(userId) != users.end();
    }
};

File: golang/355.DesignTwitter.go
const MostRecentTweetNum = 10

type Tweet struct {
	userId    int
	tweetId   int
	timestamp int
}

type User struct {
	id        int
	tweets    []*Tweet
	following map[*User]bool
}

type Twitter struct {
	timestamp int
	users     map[int]*User
}

type NewsFeed []*Tweet

func (newsFeed NewsFeed) Len() int {
	return len(newsFeed)
}

func (newsFeed NewsFeed) Less(i, j int) bool {
	return newsFeed[i].timestamp < newsFeed[j].timestamp
}

func (newsFeed NewsFeed) Swap(i, j int) {
	newsFeed[i], newsFeed[j] = newsFeed[j], newsFeed[i]
}

func (newsFeed *NewsFeed) Push(x interface{}) {
	*newsFeed = append(*newsFeed, x.(*Tweet))
}

func (newsFeed *NewsFeed) Pop() interface{} {
	n := len(*newsFeed)
	item := (*newsFeed)[n-1]
	(*newsFeed)[n-1] = nil
	*newsFeed = (*newsFeed)[0 : n-1]
	return item
}

func (this *Twitter) getUser(userId int) *User {
	if user := this.users[userId]; user != nil {
		return user
	}

	var user User
	user.id = userId
	user.following = make(map[*User]bool)
    user.following[&user] = true
	this.users[userId] = &user
	return &user
}

func (user *User) addTweet(tweet *Tweet) {
	user.tweets = append(user.tweets, tweet)
}

func (user *User) follow(followee *User) {
	user.following[followee] = true
}

func (user *User) unfollow(followee *User) {
	delete(user.following, followee)
}

/** Initialize your data structure here. */
func Constructor() Twitter {
	var twitter Twitter
	twitter.users = make(map[int]*User, 0)
	return twitter
}

/** Compose a new tweet. */
func (this *Twitter) PostTweet(userId int, tweetId int) {
	tweet := &Tweet{userId, tweetId, this.timestamp}
	this.timestamp++

	this.getUser(userId).addTweet(tweet)
}

/** Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent. */
func (this *Twitter) GetNewsFeed(userId int) []int {
	var user *User
	if user = this.users[userId]; user == nil {
		return []int{}
	}

	var newsFeed NewsFeed
	heap.Init(&newsFeed)
	for followee := range user.following {
		for i := len(followee.tweets) - 1; i >= 0 && i >= len(followee.tweets)-MostRecentTweetNum; i-- {
			currentTweet := followee.tweets[i]
			if newsFeed.Len() < MostRecentTweetNum {
				heap.Push(&newsFeed, currentTweet)
				continue
			}
			leastRecentTweet := newsFeed[0]
			if currentTweet.timestamp > leastRecentTweet.timestamp {
				heap.Pop(&newsFeed)
				heap.Push(&newsFeed, currentTweet)
			}
		}
	}

	n := newsFeed.Len()
	result := make([]int, n)
	for newsFeed.Len() > 0 {
        n--
        result[n] = heap.Pop(&newsFeed).(*Tweet).tweetId
	}
	return result
}

/** Follower follows a followee. If the operation is invalid, it should be a no-op. */
func (this *Twitter) Follow(followerId int, followeeId int) {
	follower := this.getUser(followerId)
	followee := this.getUser(followeeId)
	follower.follow(followee)
}

/** Follower unfollows a followee. If the operation is invalid, it should be a no-op. */
func (this *Twitter) Unfollow(followerId int, followeeId int) {
    if followerId == followeeId {
        return
    }
	if this.users[followerId] == nil || this.users[followeeId] == nil {
		return
	}
	follower := this.getUser(followerId)
	followee := this.getUser(followeeId)
	follower.unfollow(followee)
}

